lib/ex_machina/ecto_strategy.ex

```diff
@@ -0,0 +1,157 @@
+if Code.ensure_compiled(ExMachina) == {:module, ExMachina} do
+  defmodule PolymorphicEmbed.ExMachina.EctoStrategy do
+    @moduledoc false
+
+    # Copied from https://github.com/thoughtbot/ex_machina/blob/b2f47a36b84fded6c37434bbf9041b33b30387e9/lib/ex_machina/ecto_strategy.ex
+
+    use ExMachina.Strategy, function_name: :insert
+
+    def handle_insert(%{__meta__: %{__struct__: Ecto.Schema.Metadata}} = record, %{repo: repo}) do
+      record
+      |> cast
+      |> repo.insert!
+    end
+
+    def handle_insert(resource, options),
+      do: ExMachina.EctoStrategy.handle_insert(resource, options)
+
+    defp cast(record) do
+      record
+      |> cast_all_fields
+      |> cast_all_embeds
+      ## Change Start
+      |> cast_polymorphic_embeds
+      ## Change End
+      |> cast_all_assocs
+    end
+
+    defp cast_all_fields(%{__struct__: schema} = struct) do
+      schema
+      |> schema_fields()
+      |> Enum.reduce(struct, fn field_key, struct ->
+        casted_value = cast_field(field_key, struct)
+
+        Map.put(struct, field_key, casted_value)
+      end)
+    end
+
+    defp cast_field(field_key, %{__struct__: schema} = struct) do
+      field_type = schema.__schema__(:type, field_key)
+      value = Map.get(struct, field_key)
+
+      cast_value(field_type, value, struct)
+    end
+
+    defp cast_value(field_type, value, struct) do
+      case Ecto.Type.cast(field_type, value) do
+        {:ok, value} ->
+          value
+
+        _ ->
+          raise "Failed to cast `#{inspect(value)}` of type #{inspect(field_type)} in #{
+                  inspect(struct)
+                }."
+      end
+    end
+
+    defp cast_all_embeds(%{__struct__: schema} = struct) do
+      schema
+      |> schema_embeds()
+      |> Enum.reduce(struct, fn embed_key, struct ->
+        casted_value = struct |> Map.get(embed_key) |> cast_embed(embed_key, struct)
+
+        Map.put(struct, embed_key, casted_value)
+      end)
+    end
+
+    defp cast_embed(embeds_many, embed_key, struct) when is_list(embeds_many) do
+      Enum.map(embeds_many, &cast_embed(&1, embed_key, struct))
+    end
+
+    defp cast_embed(embed, embed_key, %{__struct__: schema}) do
+      if embed do
+        embedding_reflection = schema.__schema__(:embed, embed_key)
+        embed_type = embedding_reflection.related
+        embed_type |> struct() |> Map.merge(embed) |> cast()
+      end
+    end
+
+    ## Change Start
+    defp cast_polymorphic_embeds(%{__struct__: schema} = struct) do
+      schema
+      |> schema_polymorphic_embeds()
+      |> Enum.reduce(struct, fn embed_key, struct ->
+        casted_value = struct |> Map.get(embed_key)
+
+        Map.put(struct, embed_key, casted_value)
+      end)
+    end
```

Comment:
This won't handle nested embeds, or auto-generated binary ids on the embed schema. Making this change does the trick:
```suggestion
defp cast_polymorphic_embeds(%{__struct__: schema} = struct) do
  schema
  |> schema_polymorphic_embeds()
  |> Enum.reduce(struct, &cast_polymorphic_embed/2)
end

defp cast_polymorphic_embed(embed_key, struct) do
  casted_value =
    case Map.get(struct, embed_key) do
      embeds when is_list(embeds) -> Enum.map(embeds, &build_polymorphic_embed/1)
      embed when is_map(embed) -> build_polymorphic_embed(embed)
      embed -> embed
    end

  Map.put(struct, embed_key, casted_value)
end

defp build_polymorphic_embed(%{__struct__: schema} = struct) do
  case schema.__schema__(:autogenerate_id) do
    {key, _source, :binary_id} -> Map.put(struct, key, Ecto.UUID.generate())
    _ -> struct
  end
  |> cast()
end
```